src/main/java/st/ratpack/auth/CachingTokenValidator.java

```diff
@@ -22,38 +22,25 @@ public CachingTokenValidator(TokenValidator upstreamValidator) {
 			.maximumSize(10000L)
 			.expireAfterWrite(5L, TimeUnit.MINUTES)
 			//			.executor(Execution.current().getEventLoop())  Don't do this it makes Ratpack hang.
-			.build();
+			.build(this::loadCache);
 	}
 
 	@Override
 	public Promise<Optional<OAuthToken>> validate(String token) {
-		Promise<Optional<OAuthToken>> promiseOAuthToken = cache.getIfPresent(token);
-
-		if (promiseOAuthToken == null) {
-			//Cache miss use the upstream validator
-			logger.trace("Cache MISS: " + token);
-			promiseOAuthToken = upstreamValidator.validate(token);
-
-			//Make sure we only call the validate on the upstream once
-			promiseOAuthToken = promiseOAuthToken.cache();
-
-			promiseOAuthToken.then(optionalOAuth -> {
-				logger.trace("PUTTING: " + optionalOAuth);
-				cache.put(token, Promise.value(optionalOAuth));
-			});
-
-			//TODO The wiretap is cleaner but doesn't currently seem to work due to upstream bug
-			//Wiretap the promise so we can cache any successful results
-			//			promiseOAuthToken.wiretap(optionalResult -> {
-			//				logger.trace("Wiretap called");
-			//				if (optionalResult.isSuccess()) {
-			//					logger.trace("Caching promise value");
-			//					cache.put(token, Promise.value(optionalResult.getValue()));
-			//				}
-			//			});
-		} else {
-			logger.trace("CACHE HIT: " + token);
-		}
+		return cache.get(token);
+	}
+
+	private Promise<Optional<OAuthToken>> loadCache(String token) {
+		logger.trace("Cache MISS: {}", token);
+		//Make sure we only call the validate on the upstream once
+		Promise<Optional<OAuthToken>>
+				promiseOAuthToken = upstreamValidator.validate(token).cache();
+
+		promiseOAuthToken
+				.onError(e -> cache.invalidate(token))
+				.map(o -> o.orElse(null))
+				.onNull(() -> cache.invalidate(token))
```

Comment:
Execution of the Promise won't begin until `Promise#then` is invoked. Even then, creating promises and calling `onError`, `map`, etc creates additional execution segments that get queued to run in order so that execution order is guaranteed. So in this example

```
private Promise<Optional<OAuthToken>> loadCache(String token) {
		logger.trace("Cache MISS: {}", token);
		//Make sure we only call the validate on the upstream once
		Promise<Optional<OAuthToken>>
				promiseOAuthToken = upstreamValidator.validate(token).cache();

		// 1
		promiseOAuthToken
				.onError(e -> cache.invalidate(token))
				.map(o -> o.orElse(null))
				.onNull(() -> cache.invalidate(token))
				.then(o -> {
					// 2
					logger.trace("PUTTING: {}", o);
				});
		// 3
		try {
			return promiseOAuthToken;
		} finally {
			// 4
		}

	}
```

Execution order is 1, 3, 4, 2 so the execution of the `then` segment doesn't run until after the method exits.